File: linked_list/linked_list.py
#!/bin/python
import random

"""
Linked List implementation and functions that helps to test the code:
"""

RANDOM_LIMIT = 100
N = 10

class LinkedList:

    def __init__(self, value):
        self.value = value
        self.next = None
        self.visited = False


    def add(self, value):
        node = self
        while node.next:
            node = node.next
        node.next = LinkedList(value)


def visit_list(root):
    node = root
    while node and not node.visited:
        node.visited = True
        node = node.next


def clear_list(root):
    node = root
    while node and node.visited:
        node.visited = False
        node = node.next


def create_list(n):
    value = random.randint(0, RANDOM_LIMIT)
    root = LinkedList(value)
    node = root
    for _ in xrange(n - 1):
        value = random.randint(0, RANDOM_LIMIT)
        node.next = LinkedList(value)
        node = node.next
    return root


def create_list_2(l):
    root = None
    if l:
        root = LinkedList(l[0])
        node = root
        for i in xrange(1, len(l)):
            node.next = LinkedList(l[i])
            node = node.next
    return root


def print_list(root):
    node = root
    str_out = ""
    while node:
        str_out += "{0} ".format(node.value)
        node = node.next
    return str_out


def get_random_middle(root, n):
    if n <= 2:
        return None
    m = random.randint(0, n - 3)
    node = root.next
    for _ in xrange(m):
        node = node.next
    return node


def get_last(root):
    if not root:
        return root
    node = root
    while node.next:
        node = node.next
    return node


def print_separator():
    print "==================================================================="

File: cracking_the_coding_interview_book/chapter_2/LinkedList.cpp
#include <iostream>

using namespace std;

class Node {
    public:
        int data;
        Node* next;

        Node(int data) {
            this->data = data;
        }

        void print() {
            Node* current = this;
            while (current != NULL) {
                cout << current->data << " ";
                current = current->next;
            }
            cout << endl;
        }
};

class LinkedList {
    private:
        int n;
    public:
        Node* root;

        void insert(int data) {
            if (root == NULL) {
                root = new Node(data);
            } else {
                Node* aux = root;
                root = new Node(data);
                root->next = aux;
            }
            n++;
        }

        int size() {
            return n;
        }

        void print() {
            if (root != NULL) {
                root->print();
            }
        }

        Node* search(int data) {
            if (root == NULL)
                return NULL;
            Node* current = root;
            while (current != NULL) {
                if (current->data == data)
                    return current;
                current = current->next;
            }
            return NULL;
        }

};
